### PR TITLE
Forward port 3.next into 4.x.

### PR DESCRIPTION
**Pull request by dereuromark**
merges 3.next => 4.x.

### CODE REVIEW

File: src/Validation/Validation.php [hunk 1]
```diff
@@ -330,12 +330,12 @@ public static function comparison($check1, string $operator, $check2): bool
                 }
                 break;
             case static::COMPARE_EQUAL:
-                if ($check1 === $check2) {
+                if ($check1 == $check2) {
```
Comment (dereuromark) [inline src/Validation/Validation.php]: How did that happen?

File: src/Datasource/EntityTrait.php [hunk 2]
```diff
@@ -434,7 +434,7 @@ public function hasValue(string $field): bool
      * $entity->unset(['name', 'last_name']);
      * ```
      *
-     * @param string|array $field The field to unset.
+     * @param string|string[] $field The field to unset.
```
Comment (ADmad) [inline src/Datasource/EntityTrait.php]: This change needs to be done for `EntityInterface::unset()` too.

File: src/Utility/CookieCryptTrait.php [hunk 3]
```diff
@@ -92,10 +92,10 @@ protected function _checkCipher(string $encrypt): void
     /**
      * Decrypts $value using public $type method in Security class
      *
-     * @param array|string $values Values to decrypt
+     * @param string[]|string $values Values to decrypt
      * @param string|false $mode Encryption mode
      * @param string|null $key Used as the security salt if specified.
-     * @return string|array Decrypted values
+     * @return string|string[] Decrypted values
```
Comment (ADmad) [inline src/Utility/CookieCryptTrait.php]: Same need to be done for return type of  `_decode()` method called by `_decrypt()` else psalm is unhappy.

File: src/Validation/Validation.php (hunk 1 above)
Comment (markstory) [reply src/Validation/Validation.php]: I think it likely happened a long time ago when we changed all the `==` to `===` and didn't notice this.

**Pull request merged**